File: project_euler/Problem021/Problem021.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem021() {
    target := 10000
    divisors := make([]int, target)
    for i := 0; i < target; i++ {
        divisors[i] = -1
    }
    for i := 2; i <= target; i++ {
        getDivisorSummation(i, divisors, target)
    }
    for i := 1; i < target; i++ {
        divisors[i] -= i + 1
    }

    result := 0
    for i := 2; i <= target; i++ {
        divisor := divisors[i - 1]
        if divisor > i && divisor <= target && divisors[divisor - 1] == i {
            result += i + divisor
        }

    }
    fmt.Println(result)
}

func getDivisorSummation(value int, divisors []int, length int) {
    if -1 == divisors[value - 1] {
        index := value
        count := 1
        for index <= length {
            divisors[index - 1] = (int(math.Pow(float64(value), float64(count + 1))) - 1) / (value - 1)
            index *= value
            count++
        }

        for i := int(math.Min(float64(value - 1), float64(length / value))); i >= 2; i-- {
            divisors[value * i - 1] = divisors[value - 1] * divisors[i - 1]
        }
    } else {
        for i := int(math.Min(float64(value - 1), float64(length / value))); i >= 2; i-- {
            if isRelativelyPrime(value, i) && -1 == divisors[value * i - 1] {
                divisors[value * i - 1] = divisors[value - 1] * divisors[i - 1]
            }
        }
    }
}

func isRelativelyPrime(i int, j int) bool {
    for 0 != i % j {
        i, j = j, i % j
    }
    return 1 == j
}

File: project_euler/Problem021/Problem021.py
import math
from typing import List

class Problem021:
    def solution(self) -> None:
        target = 10000
        divisors = [-1 for _ in range(target)]
        for i in range(2, target + 1):
            self.__get_divisor_summation(i, divisors, target)
        for i in range(1, target):
            divisors[i] -= i + 1

        result = 0
        for i in range(2, target + 1):
            divisor = divisors[i - 1]
            if divisor > i and divisor <= target and divisors[divisor - 1] == i:
                result += i + divisor
        print(result)

    def __get_divisor_summation(self, value: int, divisors: List[int], length: int) -> None:
        if -1 == divisors[value - 1]:
            index, count = value, 1
            while index <= length:
                divisors[index - 1] = (int(math.pow(value, count + 1)) - 1) // (value - 1)
                index *= value
                count += 1

            for i in range(min(value - 1, length // value), 1, -1):
                divisors[value * i - 1] = divisors[value - 1] * divisors[i - 1]
        else:
            for i in range(min(value - 1, length // value), 1, -1):
                if self.__is_relatively_prime(value, i) and -1 == divisors[value * i - 1]:
                    divisors[value * i - 1] = divisors[value - 1] * divisors[i - 1]

    def __is_relatively_prime(self, i: int, j: int) -> bool:
        while i % j:
            i, j = j, i % j
        return 1 == j
